### PR TITLE
feat(sdk): Remove clones of large responses in Sliding Sync

### PR DESCRIPTION
**Pull request by Hywan**
I admit this patch is quite tricky. Please try to follow me.

So, first off, in `SlidingSyncRoom::new`, we were clearing the timeline, because somehow it exists twice in memory at this step.

Which led me to understand how `SlidingSync::handle_response` was working. I've clarified how this part of the code works. We are dealing with 2 kind of responses for a specific reason: `SyncResponse` and `v4::Response`, now it's documented and I hope it's clearer.

Then, I notice that we were passing a clone of the entire sliding sync response (`v4::Response`) to `Client::process_sliding_sync`. I thought it was suboptimal, so I've updated the code to take a reference. It led me to update `BaseClient::process_sliding_sync`. It was a little bit tricky, but I reckon we have less clones now than before.

And now, back to `SlidingSync::handle_response`, I was able to compute the `timeline` correctly by draining it from the `v4::Response`, or by moving it from `SyncResponse`. So it's no longer necessary to have this clearing code inside `SlidingSyncRoom::new`. Honestly it has nothing to do at this place before.

To conclude: We have cleaner code, and less clones.

What thing I reckon could be optimized, is that the entire `timeline` (`Vec<TimelineEvent>`) is cloned to be passed to `Client::handle_timeline`. So this timeline exists in 2 places: in Sliding Sync, and somewhere else. I don't believe it's a problem now, that's how it works, but we must be aware of that.

### CODE REVIEW

File: crates/matrix-sdk-base/src/sliding_sync.rs [hunk 1]
```diff
@@ -49,21 +50,31 @@ impl BaseClient {
...
+        let to_device_events = to_device.as_ref().map(|v4| v4.events.clone()).unwrap_or_default();
 
         // Destructure the single `None` of the E2EE extension into separate objects
-        // since that's what the OlmMachine API expects. Passing in the default
-        // empty maps and vecs for this is completely fine, since the OlmMachine
+        // since that's what the `OlmMachine` API expects. Passing in the default
+        // empty maps and vecs for this is completely fine, since the `OlmMachine`
         // assumes empty maps/vecs mean no change in the one-time key counts.
+
+        // We declare default values that can be referenced hereinbelow. When we try to
```
Comment (jplatte) [inline crates/matrix-sdk-base/src/sliding_sync.rs:60]: Is "hereinbelow" really a word that exists? 😄

File: crates/matrix-sdk-base/src/sliding_sync.rs [hunk 2]
```diff
@@ -49,21 +50,31 @@ impl BaseClient {
...
+        // extract values from `e2ee`, that would be unfortunate to clone the
+        // value just to pass them (to remove them `e2ee`) as a reference later.
+        let device_one_time_keys_count = Default::default();
+        let device_unused_fallback_key_types = Default::default();
```
Comment (jplatte) [inline crates/matrix-sdk-base/src/sliding_sync.rs:64]: `Type::default` would make this easier to read IMHO.

File: crates/matrix-sdk-base/src/sliding_sync.rs [hunk 3]
```diff
@@ -49,21 +50,31 @@ impl BaseClient {
...
-                    e2ee.device_one_time_keys_count,
-                    e2ee.device_unused_fallback_key_types,
+                    e2ee.device_lists.clone(),
+                    &e2ee.device_one_time_keys_count,
+                    &e2ee.device_unused_fallback_key_types,
                 )
             })
-            .unwrap_or_default();
+            .unwrap_or_else(|| {
+                (Default::default(), &device_one_time_keys_count, &device_unused_fallback_key_types)
```
Comment (jplatte) [inline crates/matrix-sdk-base/src/sliding_sync.rs:76]: Same here.

File: crates/matrix-sdk/src/sliding_sync/mod.rs [hunk 4]
```diff
@@ -975,28 +975,45 @@ impl SlidingSync {
...
 
+    /// Handle the HTTP response.
+    ///
+    /// But which response? `v4::Response`, aka the Sliding Sync response, or
+    /// `SyncResponse` which still relies on `v3`?
+    /// Well that's tricky. We have both here, because this Sliding Sync
+    /// implementation is still experimental, and we didn't want to be too
+    /// invasive. Thus, `SyncResponse` doesn't support Sliding Sync yet. Hence
+    /// the fact this method handles both at the same time. It's not super
+    /// annoying but it was important to clarify that.
```
Comment (jplatte) [inline crates/matrix-sdk/src/sliding_sync/mod.rs:986]: I remember we have some code path that's common between the two sync types, but this fn is in sliding_sync so it can't be used from elsewhere. This comment adds confusion when it should be clarifying things 😅

File: crates/matrix-sdk/src/sliding_sync/room.rs [hunk 5]
```diff
@@ -41,12 +41,9 @@ impl SlidingSyncRoom {
     pub(super) fn new(
         client: Client,
         room_id: OwnedRoomId,
-        mut inner: v4::SlidingSyncRoom,
+        inner: v4::SlidingSyncRoom,
         timeline: Vec<SyncTimelineEvent>,
     ) -> Self {
-        // we overwrite to only keep one copy
-        inner.timeline = vec![];
```
Comment (jplatte) [inline crates/matrix-sdk/src/sliding_sync/room.rs:48]: Why was this removed? It's not clear.

File: crates/matrix-sdk-base/src/sliding_sync.rs (hunk 1 above)
Comment (Hywan) [reply crates/matrix-sdk-base/src/sliding_sync.rs:60]: Yes, and it's rare to use it, so please, don't keep my pleasure away :-p.

File: crates/matrix-sdk/src/sliding_sync/room.rs (hunk 5 above)
Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/room.rs:48]: Because it's not necessary anymore (I guess it wasn't necessary before, but now I'm sure).

File: crates/matrix-sdk-base/src/sliding_sync.rs (hunk 2 above)
Comment (Hywan) [reply crates/matrix-sdk-base/src/sliding_sync.rs:64]: Done.

File: crates/matrix-sdk/src/sliding_sync/mod.rs (hunk 4 above)
Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/mod.rs]: Updated.

File: crates/matrix-sdk-base/src/sliding_sync.rs [hunk 6]
```diff
@@ -60,8 +64,8 @@ impl BaseClient {
         // We declare default values that can be referenced hereinbelow. When we try to
         // extract values from `e2ee`, that would be unfortunate to clone the
         // value just to pass them (to remove them `e2ee`) as a reference later.
-        let device_one_time_keys_count = Default::default();
-        let device_unused_fallback_key_types = Default::default();
+        let device_one_time_keys_count = BTreeMap::<DeviceKeyAlgorithm, UInt>::default();
```
Comment (jplatte) [inline crates/matrix-sdk-base/src/sliding_sync.rs:67]: I just meant the `BTreeMap`, not `::<_, _>`. But I guess it doesn't hurt either.

File: crates/matrix-sdk/src/sliding_sync/mod.rs [hunk 7]
```diff
@@ -978,12 +978,8 @@ impl SlidingSync {
...
     ///
     /// But which response? `v4::Response`, aka the Sliding Sync response, or
-    /// `SyncResponse` which still relies on `v3`?
-    /// Well that's tricky. We have both here, because this Sliding Sync
-    /// implementation is still experimental, and we didn't want to be too
-    /// invasive. Thus, `SyncResponse` doesn't support Sliding Sync yet. Hence
-    /// the fact this method handles both at the same time. It's not super
-    /// annoying but it was important to clarify that.
+    /// `SyncResponse`? We have both because `SyncResponse` doesn't support
+    /// Sliding Sync yet.
```
Comment (jplatte) [inline crates/matrix-sdk/src/sliding_sync/mod.rs:982]: What does "we have both" mean? This function takes a `v4::Response`, does it convert it?

Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/mod.rs:982]: Oh, when it processes `v4::Response` just below, it gets a `SyncResponse`. It doesn't receive it as an argument, that's true, but the method works with both responses.

Comment (jplatte) [reply crates/matrix-sdk/src/sliding_sync/mod.rs:982]: Okay. Well I think it should be clear though, `SyncResponse` is not `v3::Response` 😅 

So I'm still not convinced about this comment..

Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/mod.rs:982]: I'll rewrite it Monday :-).

Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/mod.rs]: Done :-).

File: crates/matrix-sdk/src/sliding_sync/mod.rs [hunk 8]
```diff
@@ -976,17 +976,15 @@ impl SlidingSync {
...
-    /// `SyncResponse`? We have both because `SyncResponse` doesn't support
-    /// Sliding Sync yet.
     #[instrument(skip_all, fields(views = views.len()))]
     async fn handle_response(
         &self,
         sliding_sync_response: v4::Response,
         extensions: Option<ExtensionsConfig>,
         views: &mut BTreeMap<String, SlidingSyncViewRequestGenerator>,
     ) -> Result<UpdateSummary, crate::Error> {
+        // Handle and transform a Sliding Sync Response to a `SyncResponse`.
```
Comment (jplatte) [inline crates/matrix-sdk/src/sliding_sync/mod.rs:986]: 👍🏼

**Review by jplatte — APPROVED** [review]
